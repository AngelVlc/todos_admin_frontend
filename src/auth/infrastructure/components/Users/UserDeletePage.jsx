import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link, useParams, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import {
  GetUserByIdUseCase,
  DeleteUserByIdUseCase,
} from "../../../application/users";

export const UserDeletePage = () => {
  const [user, setUser] = useState(null);
  const { useCaseFactory } = useContext(AppContext);
  let { userId } = useParams();
  let history = useHistory();

  const getUser = useCallback(async () => {
    const getUserByIdUseCase = useCaseFactory.get(GetUserByIdUseCase);
    const user = await getUserByIdUseCase.execute(userId);

    setUser(user);
  }, [userId, useCaseFactory]);

  useEffect(() => {
    getUser();
  }, [userId, getUser]);

  const deleteUser = async () => {
    const deleteUserByIdUseCase = useCaseFactory.get(DeleteUserByIdUseCase);

    if (await deleteUserByIdUseCase.execute(userId)) {
      history.push("/users");
    }
  };

  return (
    <>
      {user && (
        <div className="container">
          <h3 className="title">
            {user.isAdmin
              ? `Delete admin user '${user.name}'?`
              : `Delete user ${user.name}?`}
          </h3>
          <nav className="breadcrumb" aria-label="breadcrumbs">
            <ul>
              <li>
                <Link to={`/`}>Home</Link>
              </li>
              <li>
                <Link to={`/users`}>Users</Link>
              </li>
              <li>
                <Link to={`/users`}>Users</Link>
              </li>
              <li>
                <Link aria-current="page" to={`/users/${userId}/edit`}>
                  {user.name}
                </Link>
              </li>
              <li className="is-active">
                <Link
                  aria-current="page"
                  to={`/users/${userId}/delete`}
                >{`Delete ${user.name}`}</Link>
              </li>
            </ul>
          </nav>
          <div className="buttons">
            <button
              className="button is-danger"
              data-testid="yes"
              onClick={() => deleteUser()}
            >
              YES
            </button>
            <button
              className="button"
              data-testid="no"
              onClick={() => history.goBack()}
            >
              NO
            </button>
          </div>
        </div>
      )}
    </>
  );
};
