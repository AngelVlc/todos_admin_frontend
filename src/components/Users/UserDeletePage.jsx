import React, { useEffect, useState } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import axios from 'axios';

export const UserDeletePage = () => {
  const [user, setUser] = useState(null);
  let { userId } = useParams();
  let history = useHistory();

  useEffect(() => {
    const getUser = async () => {
      const res = await axios.get(`users/${userId}`);
      let title = `Delete user ${res.data.name}`;
      if (res.data.isAdmin) {
        title = `Delete admin user '${res.data.name}'?`;
      }
      const info = {
        ...res.data,
        title: title
      };
      setUser(info);
    }
    getUser();
  }, [userId]);

  const deleteUser = async () => {
    await axios.delete(`users/${userId}`);
    history.push('/users');
  }

  return (
    <>
      {user &&
        <div className="container">
          <h3 className="title">{user.title}</h3>
          <div className="buttons">
            <button className="button is-danger" data-testid="yes" onClick={() => deleteUser()}>YES</button>
            <button className="button" data-testid="no" onClick={() => history.goBack()}>NO</button>
          </div>
        </div>
      }
    </>
  );
}
