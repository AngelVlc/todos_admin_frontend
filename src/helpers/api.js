import { requestPending, requestDone, requestFailed } from '../actions'

const backendUrl = process.env.REACT_APP_BACKEND_URL || 'http://localhost:5001'

export const doGetToken = (loginDto, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(loginDto)
        };
        fetch(`${backendUrl}/auth/token`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (!res.ok) {
                    res.text().then(txt => reject(txt));
                    return
                }
                res.json().then(obj => {
                    resolve(obj);
                })
            })
            .catch(error => {
                requestsDispatch(requestDone());
                reject(error.message);
            })
    });
}

const getHeaders = (method, token) => {
    return {
        method: method,
        headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` },
    };
}

export const doGet = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('GET', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    res.json().then(obj => {
                        resolve(obj);
                    })
                } else {
                    res.text().then(txt => {
                        requestsDispatch(requestFailed(txt));
                        reject(txt)
                    });
                }
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doDelete = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('DELETE', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    resolve();
                    return
                }
                res.text().then(txt => {
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doPost = (endpoint, body, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('POST', token);
        requestOptions.body = JSON.stringify(body);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                res.text().then(txt => {
                    if (res.ok) {
                        resolve(txt);
                        return
                    }
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}
