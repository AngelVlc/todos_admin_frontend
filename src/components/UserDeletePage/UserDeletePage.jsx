import React, { useEffect, useContext, useState } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { AppContext } from '../../contexts/AppContext';
import { doGet, doDelete } from '../../helpers/api';

export const UserDeletePage = () => {
  const { auth, requestsDispatch } = useContext(AppContext)
  const [user, setUser] = useState(null);
  let { userId } = useParams();
  let history = useHistory();

  useEffect(() => {
    const getUser = async () => {
      const res = await doGet(`users/${userId}`, auth.info.token, requestsDispatch)
      let title = `Delete user ${res.name}`
      if (res.isAdmin) {
        title = `Delete admin user '${res.name}'?`
      }
      const info = {
        ...res,
        title: title
      }
      setUser(info);
    }
    if (auth.info) {
      getUser();
    }
  }, [userId, auth.info, requestsDispatch]);

  const deleteUser = async () => {
    await doDelete(`users/${userId}`, auth.info.token, requestsDispatch)
    history.goBack();
  }

  return (
    <>
      {user &&
        <div className="container">
          <h3 className="title">{user.title}</h3>
          <div className="buttons">
            <button className="button is-danger" data-testid="yes" onClick={() => deleteUser()}>YES</button>
            <button className="button" data-testid="no" onClick={() => history.goBack()}>NO</button>
          </div>
        </div>
      }
    </>
  );
}
