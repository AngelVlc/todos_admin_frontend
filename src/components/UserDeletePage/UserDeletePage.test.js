import React from 'react'
import { render, cleanup, fireEvent, wait } from '@testing-library/react'
import { UserDeletePage } from './UserDeletePage'
import { AppContext } from '../../contexts/AppContext'
import * as api from '../../helpers/api';
import { MemoryRouter, Route } from 'react-router-dom'
import { act } from 'react-dom/test-utils';

jest.mock('../../helpers/api');
const mockHistoryGoBack = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        goBack: mockHistoryGoBack
    })
}));

const renderWithContextAndRouter = (component) => {
    api.doGet.mockResolvedValue({
        id: 2,
        name: 'user',
        isAdmin: false
    });
    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/user/2/delete`]}>
                    <Route path="/user/:userId/delete">
                        {component}
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

afterEach(cleanup)

it('should match the snapshot', async () => {
    let fragment;
    await act(async () => {
        const { asFragment } = renderWithContextAndRouter(<UserDeletePage />);
        fragment = asFragment;
    });
    expect(fragment(<UserDeletePage />)).toMatchSnapshot();
});

it('should cancel the deletion', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouter(<UserDeletePage />);
    });

    await wait(() => {
        fireEvent.click(container.getByTestId('no'));
    })

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

it('should delete the user', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouter(<UserDeletePage />);
    });

    api.doDelete.mockResolvedValue({});

    await wait(() => {
        fireEvent.click(container.getByTestId('yes'));
    })

    expect(api.doDelete.mock.calls.length).toBe(1);
    expect(api.doDelete.mock.calls[0][0]).toBe('users/2');
    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});
