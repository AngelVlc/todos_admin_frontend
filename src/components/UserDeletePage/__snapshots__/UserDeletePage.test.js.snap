// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <h4>
    Delete user 'user' (is admin: false)?
  </h4>
  <button
    data-testid="yes"
  >
    YES
  </button>
  <button
    data-testid="no"
  >
    NO
  </button>
</DocumentFragment>
`;
